lalrpop-util/src/state_machine.rs

```diff
@@ -0,0 +1,671 @@
+    /// current lookahead that is triggering the reduction (it is
+    /// `None` for EOF).
+    ///
+    /// The `states` and `symbols` vectors represent the internal
+    /// state machine vectors; they are given to `reduce` so that it
+    /// can pop off states that no longer apply (and consume their
+    /// symbols). At the end, it should also push the new state and
+    /// symbol produced.
+    ///
+    /// Returns a `Some` if we reduced the start state and hence
+    /// parsing is complete, or if we encountered an irrecoverable
+    /// error.
+    ///
+    /// FIXME. It would be nice to not have so much logic live in
+    /// reduce.  It should just be given an iterator of popped symbols
+    /// and return the newly produced symbol (or error). We can use
+    /// `simulate_reduce` and our own information to drive the rest,
+    /// right? This would also allow us -- I think -- to extend error
+    /// recovery to cover user-produced errors.
+    fn reduce(
+        &mut self,
+        reduce_index: Self::ReduceIndex,
+        start_location: Option<&Self::Location>,
+        states: &mut Vec<Self::StateIndex>,
+        symbols: &mut Vec<SymbolTriple<Self>>,
+    ) -> Option<ParseResult<Self>>;
+
+    /// Returns information about how many states will be popped
+    /// during a reduction, and what nonterminal would be produced as
+    /// a result.
+    fn simulate_reduce(&self, action: Self::ReduceIndex) -> SimulatedReduce<Self>;
+}
+
+pub trait ParserAction<D: ParserDefinition>: Copy + Clone + Debug {
+    fn as_shift(self) -> Option<D::StateIndex>;
+    fn as_reduce(self) -> Option<D::ReduceIndex>;
+    fn is_shift(self) -> bool;
+    fn is_reduce(self) -> bool;
+    fn is_error(self) -> bool;
+}
+
+pub enum SimulatedReduce<D: ParserDefinition> {
+    Reduce {
+        states_to_pop: usize,
+        nonterminal_produced: D::NonterminalIndex,
+    },
+
+    // This reduce is the "start" fn, so the parse is done.
+    Accept,
+}
+
+// These aliases are an elaborate hack to get around
+// the warnings when you define a type alias like `type Foo<D: Trait>`
+#[doc(hidden)]
+pub type Location<D> = <D as ParserDefinition>::Location;
+#[doc(hidden)]
+pub type Token<D> = <D as ParserDefinition>::Token;
+#[doc(hidden)]
+pub type Error<D> = <D as ParserDefinition>::Error;
+#[doc(hidden)]
+pub type Success<D> = <D as ParserDefinition>::Success;
+#[doc(hidden)]
+pub type Symbol<D> = <D as ParserDefinition>::Symbol;
+
+pub type ParseError<D> = ::ParseError<Location<D>, Token<D>, Error<D>>;
+pub type ParseResult<D> = Result<Success<D>, ParseError<D>>;
+pub type TokenTriple<D> = (Location<D>, Token<D>, Location<D>);
+pub type SymbolTriple<D> = (Location<D>, Symbol<D>, Location<D>);
+pub type ErrorRecovery<D> = ::ErrorRecovery<Location<D>, Token<D>, Error<D>>;
+
+pub struct Parser<D, I>
+where
+    D: ParserDefinition,
+    I: Iterator<Item = Result<TokenTriple<D>, ParseError<D>>>,
+{
+    definition: D,
+    tokens: I,
```

Comment:
If `tokens` were taken as an argument instead of being part of the parser itself we could reset and reuse the parser over multiple inputs. I might do that as a followup if it sounds reasonable.

Reply:
Sure, that'd be fine. I guess there is a little bit of state we could reuse (e.g., vectors for stacks and things)